votor/src/event_handler.rs

```diff
@@ -387,6 +435,61 @@ impl EventHandler {
         Ok(votes)
     }
 
+    /// Under normal cases we should have a parent ready for first slot of every window.
+    /// But it could be we joined when the later slots of the window are finalized, then
+    /// we never saw the parent ready for the first slot and haven't voted for first slot
+    /// so we can't keep processing rest of the window. This is especially a problem for
+    /// cluster standstill.
+    /// If we get a finalization cert for later slots of the window and we have the block
+    /// replayed, trace back to the first slot of the window and emit parent ready.
+    fn add_missing_parent_ready(
+        block: Block,
+        ctx: &SharedContext,
+        vctx: &mut VotingContext,
+        local_context: &mut LocalContext,
+    ) -> Option<(Slot, Block)> {
+        let (slot, block_id) = block;
+        let first_slot_of_window = first_of_consecutive_leader_slots(slot);
+        if first_slot_of_window == slot || first_slot_of_window == 0 {
+            // No need to trigger parent ready for the first slot of the window
+            return None;
+        }
+        if vctx.vote_history.highest_parent_ready_slot() >= Some(first_slot_of_window)
+            || !local_context.finalized_blocks.contains(&block)
+        {
+            return None;
+        }
+        // If the block is missing, we can't trigger parent ready
+        let bank_forks_r = ctx.bank_forks.read().unwrap();
+        let bank = bank_forks_r.get(slot)?;
+        let first_slot_of_window_bank = bank_forks_r.get(first_slot_of_window)?;
+        drop(bank_forks_r);
+        if !bank.is_frozen() {
+            // We haven't finished replay for the block, so we can't trigger parent ready
+            return None;
+        }
+        if bank.block_id() != Some(block_id) {
```

Comment:
Pointing this out as a sanity check, I believe it is true that a block within the leader window being finalized implies that the first slot in the window is also finalized (due to the consecutive check in tryNotar).
Thus it's impossible for this block_id to match, but first_slot_of_window_bank to not be an ancestor of bank.

Reply:
@AshwinSekar I changed ParentReady to trigger on the finalized block instead of the first slot of the window, to be consistent with restarting on a slot in the middle, what do you think? Tests look happy.

Reply:
sorry I missed this, I believe this should be fine. The key is that we'll be able to vote skip and notarize-fallback if needed.